### ISSUE TITLE
Make cargo global to be usable by any users on the system

### ISSUE THREAD

**Issue by allenzhao64** [rust-lang/cargo#8518]
**Describe the problem you are trying to solve**
I am trying to install rust/cargo to be usable by any users on the system. The current installation is per-user based, cargo is installed under user's home directory, and all the cache and registry are created/updated under the CARGO_HOME/. 

I can certainly use CARGO_HOME to install the binary to a global location, such as /opt/cargo/, however, then everyone is sharing the same registry and cache, and permission of these directories can stand in other people's way. 

Even after I play with sticky bits and POSIX ACL, I found out the dynamic download/extraction from github.com will force a file permission that only user can read/write:

Example: 
[root@linux-2018 cargo]# ls -al /opt/cargo/registry/src/github.com-1ecc6299db9ec823/term-0.6.1/src/
total 40
drwxrwsrwx+ 3 simdev root  4096 Jul 17 15:35 .
drwxrwsrwx+ 4 simdev root  4096 Jul 17 15:35 ..
-rw-------+ 1 simdev root 13428 Aug  5  2019 lib.rs
drwxrwsrwx+ 3 simdev root  4096 Jul 17 15:35 terminfo
-rw-------+ 1 simdev root 12148 Aug  5  2019 win.rs

The only way I may be able to fix it: delete the cache/registry to clean for other people. But this is still error prom. 

**Describe the solution you'd like**
maybe the registry and cache directories should prepend/append the username, such as registry-<user> and cache-<user>? This is not a fix, but only a workaround, but it will help the situation. 

**Notes**
The better way is to make cargo more globally accessible, instead of being self-centric to individual users. Maybe this feature is already there, but I just do not know?

In a big team or organization, we would like to manage/audit tools chain globally. Having developer installing their own tools is a school/amateur hack. This is just something I feel so strange when I work on the rust/cargo project.

**Comment by ehuss** [comment]
Maybe I'm a bit confused, but there seems to be two difference concepts at play here?  The installation of rustc/cargo doesn't necessarily need to go into `CARGO_HOME`. The executables can go into a global bin directory.  Most operating system package managers offer rustc and cargo packages for installation, or you can use a [standalone installer](https://forge.rust-lang.org/infra/other-installation-methods.html#standalone-installers).

As for the cache files, due to permission, I don't think it is feasible to have a globally shared cache between users (at least, not without some client/server architecture).

**Comment by allenzhao64** [comment]
I used the online installer `curl -sSf https://sh.rustup.rs | sh`, which either install into $HOME/.cargo for personal use, or $CARGO_HOME (which I can set to global location such as /opt/cargo/)

But then I ran into issue when multiple users using the same $CARGO_HOME, because registry and cache got permission issue.

I think you are hinting at not using $CARGO_HOME after initial installation (I have to set it for global installation). and I just confirm that witihout CARGO_HOME set, $HOME/.cargo is always used, that resolves the permission issue. Just test it, looks like that is the case.
